Handle remote peers not having block requested by root

### Description
When a remote peer doesn't have the block requested by root, it should return no response chunks.  Teku however is expecting it to always return a single chunk.

We should fix this to return an empty `Optional` to indicate the peer does not have the block and cleanly handle that case, rather than throwing an exception.

Note the error is logged at DEBUG level so it's not noise in logs but is incorrect behaviour.

```
{"timestamp":"2020-10-20T04:58:49,911","level":"DEBUG","thread":"nioEventLoopGroup-3-3","class":"FetchBlockTask","message":"Failed to fetch block 0xcc92333e8ab9ca670c466394c5e54f7142235b654ca485461dc104d9672b7be4","throwable":" java.util.concurrent.CompletionException: java.lang.NullPointerException: No response received when single response expected\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamImpl.completeSuccessfully(ResponseStreamImpl.java:75)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.lambda$completeRequest$0(Eth2OutgoingRequestHandler.java:196)\n\tat java.base/java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:753)\n\tat java.base/java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731)\n\tat java.base/java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2172)\n\tat tech.pegasys.teku.infrastructure.async.SafeFuture.thenAccept(SafeFuture.java:390)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.completeRequest(Eth2OutgoingRequestHandler.java:193)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.Eth2OutgoingRequestHandler.readComplete(Eth2OutgoingRequestHandler.java:157)\n\tat tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.onRemoteWriteClosed(RpcHandler.java:191)\n\tat tech.pegasys.teku.networking.p2p.libp2p.rpc.RpcHandler$Controller.userEventTriggered(RpcHandler.java:184)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:346)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:332)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:324)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.userEventTriggered(DefaultChannelPipeline.java:1428)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:346)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:332)\n\tat io.netty.channel.DefaultChannelPipeline.fireUserEventTriggered(DefaultChannelPipeline.java:913)\n\tat io.libp2p.etc.util.netty.mux.MuxChannel.onRemoteDisconnected(MuxChannel.kt:58)\n\tat io.libp2p.etc.util.netty.mux.AbstractMuxHandler.onRemoteDisconnect(AbstractMuxHandler.kt:72)\n\tat io.libp2p.mux.MuxHandler.channelRead(MuxHandler.kt:37)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)\n\tat io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)\n\tat io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)\n\tat io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)\n\tat io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)\n\tat io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:311)\n\tat io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:425)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)\n\tat io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)\n\tat io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)\n\tat io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)\n\tat io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)\n\tat io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)\n\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Thread.java:832)\nCaused by: java.lang.NullPointerException: No response received when single response expected\n\tat com.google.common.base.Preconditions.checkNotNull(Preconditions.java:897)\n\tat tech.pegasys.teku.networking.eth2.rpc.core.ResponseStreamImpl.lambda$expectSingleResponse$1(ResponseStreamImpl.java:42)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)\n\t... 58 more\n"}
```

--- a/networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java
+++ b/networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRangeIntegrationTest.java
@@ -155,7 +155,7 @@ public abstract class BeaconBlocksByRangeIntegrationTest {
     recentChainData1.updateHead(block2Root, block2.getSlot());
 
     peer1.disconnectImmediately(Optional.empty(), false);
-    final SafeFuture<SignedBeaconBlock> res = peer1.requestBlockBySlot(UInt64.ONE);
+    final SafeFuture<Optional<SignedBeaconBlock>> res = peer1.requestBlockBySlot(UInt64.ONE);
 
     waitFor(() -> assertThat(res).isDone());
     assertThat(res).isCompletedExceptionally();
@@ -171,11 +171,18 @@ public abstract class BeaconBlocksByRangeIntegrationTest {
     recentChainData1.updateHead(block2Root, block2.getSlot());
 
     Waiter.waitFor(peer1.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS));
-    final SafeFuture<SignedBeaconBlock> res = peer1.requestBlockBySlot(UInt64.ONE);
+    final SafeFuture<Optional<SignedBeaconBlock>> res = peer1.requestBlockBySlot(UInt64.ONE);
 
     waitFor(() -> assertThat(res).isDone());
     assertThat(res).isCompletedExceptionally();
     assertThatThrownBy(res::get).hasRootCauseInstanceOf(PeerDisconnectedException.class);
+  }
+
+  @Test
+  void requestBlockBySlot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception {
+    final Optional<SignedBeaconBlock> result =
+        waitFor(peer1.requestBlockBySlot(UInt64.valueOf(49382982)));
+    assertThat(result).isEmpty();
   }
 
   private List<SignedBeaconBlock> requestBlocks()

--- a/networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java
+++ b/networking/eth2/src/integration-test/java/tech/pegasys/teku/networking/eth2/BeaconBlocksByRootIntegrationTest.java
@@ -148,7 +148,7 @@ public abstract class BeaconBlocksByRootIntegrationTest {
     final Bytes32 blockHash = block.getMessage().hash_tree_root();
 
     peer1.disconnectImmediately(Optional.empty(), false);
-    final SafeFuture<SignedBeaconBlock> res = peer1.requestBlockByRoot(blockHash);
+    final SafeFuture<Optional<SignedBeaconBlock>> res = peer1.requestBlockByRoot(blockHash);
 
     waitFor(() -> assertThat(res).isDone());
     assertThat(res).isCompletedExceptionally();
@@ -162,7 +162,7 @@ public abstract class BeaconBlocksByRootIntegrationTest {
     final Bytes32 blockHash = block.getMessage().hash_tree_root();
 
     Waiter.waitFor(peer1.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS));
-    final SafeFuture<SignedBeaconBlock> res = peer1.requestBlockByRoot(blockHash);
+    final SafeFuture<Optional<SignedBeaconBlock>> res = peer1.requestBlockByRoot(blockHash);
 
     waitFor(() -> assertThat(res).isDone());
     assertThat(res).isCompletedExceptionally();
@@ -204,6 +204,13 @@ public abstract class BeaconBlocksByRootIntegrationTest {
 
     final List<SignedBeaconBlock> response = requestBlocks(blockRoots);
     assertThat(response).containsExactlyElementsOf(blocks);
+  }
+
+  @Test
+  void requestBlockByRoot_shouldReturnEmptyWhenBlockIsNotKnown() throws Exception {
+    final Optional<SignedBeaconBlock> result =
+        waitFor(peer1.requestBlockByRoot(Bytes32.fromHexStringLenient("0x123456789")));
+    assertThat(result).isEmpty();
   }
 
   private SignedBeaconBlock addBlock() {

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2Peer.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2Peer.java
@@ -178,18 +178,19 @@ public class Eth2Peer extends DelegatingPeer implements Peer, SyncSource {
     return requestStream(blockByRoot, new BeaconBlocksByRootRequestMessage(blockRoots), listener);
   }
 
-  public SafeFuture<SignedBeaconBlock> requestBlockBySlot(final UInt64 slot) {
+  public SafeFuture<Optional<SignedBeaconBlock>> requestBlockBySlot(final UInt64 slot) {
     final Eth2RpcMethod<BeaconBlocksByRangeRequestMessage, SignedBeaconBlock> blocksByRange =
         rpcMethods.beaconBlocksByRange();
     final BeaconBlocksByRangeRequestMessage request =
         new BeaconBlocksByRangeRequestMessage(slot, UInt64.ONE, UInt64.ONE);
-    return requestSingleItem(blocksByRange, request);
+    return requestOptionalItem(blocksByRange, request);
   }
 
-  public SafeFuture<SignedBeaconBlock> requestBlockByRoot(final Bytes32 blockRoot) {
+  public SafeFuture<Optional<SignedBeaconBlock>> requestBlockByRoot(final Bytes32 blockRoot) {
     final Eth2RpcMethod<BeaconBlocksByRootRequestMessage, SignedBeaconBlock> blockByRoot =
         rpcMethods.beaconBlocksByRoot();
-    return requestSingleItem(blockByRoot, new BeaconBlocksByRootRequestMessage(List.of(blockRoot)));
+    return requestOptionalItem(
+        blockByRoot, new BeaconBlocksByRootRequestMessage(List.of(blockRoot)));
   }
 
   @Override
@@ -256,6 +257,14 @@ public class Eth2Peer extends DelegatingPeer implements Peer, SyncSource {
     final Eth2OutgoingRequestHandler<I, O> handler =
         method.createOutgoingRequestHandler(request.getMaximumRequestChunks());
     SafeFuture<O> respFuture = handler.getResponseStream().expectSingleResponse();
+    return sendRequest(method, request, handler).thenCompose(__ -> respFuture);
+  }
+
+  public <I extends RpcRequest, O> SafeFuture<Optional<O>> requestOptionalItem(
+      final Eth2RpcMethod<I, O> method, final I request) {
+    final Eth2OutgoingRequestHandler<I, O> handler =
+        method.createOutgoingRequestHandler(request.getMaximumRequestChunks());
+    SafeFuture<Optional<O>> respFuture = handler.getResponseStream().expectOptionalResponse();
     return sendRequest(method, request, handler).thenCompose(__ -> respFuture);
   }
 

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidator.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidator.java
@@ -14,7 +14,6 @@
 package tech.pegasys.teku.networking.eth2.peers;
 
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.base.Throwables;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -183,13 +182,6 @@ public class PeerChainValidator {
       LOG.trace(
           "Request required checkpoint block from peer {}: {}", peer.getId(), checkpointToVerify);
       return peer.requestBlockByRoot(checkpointToVerify.getRoot())
-          .exceptionally(
-              err -> {
-                if (Throwables.getRootCause(err) instanceof NullPointerException) {
-                  return Optional.empty();
-                }
-                throw new RuntimeException(err);
-              })
           // When requesting block by root, there is no explicit guarantee that the block is
           // canonical.
           // So, double-check by requesting the block by slot to make sure the peer considers this

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidator.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidator.java
@@ -183,8 +183,6 @@ public class PeerChainValidator {
       LOG.trace(
           "Request required checkpoint block from peer {}: {}", peer.getId(), checkpointToVerify);
       return peer.requestBlockByRoot(checkpointToVerify.getRoot())
-          // Map result to an optional block
-          .thenApply(Optional::of)
           .exceptionally(
               err -> {
                 if (Throwables.getRootCause(err) instanceof NullPointerException) {
@@ -205,9 +203,11 @@ public class PeerChainValidator {
                                   .thenApply(
                                       blockBySlot -> {
                                         final boolean blockMatches =
-                                            blockBySlot
-                                                .getRoot()
-                                                .equals(checkpointToVerify.getRoot());
+                                            blockBySlot.isPresent()
+                                                && blockBySlot
+                                                    .get()
+                                                    .getRoot()
+                                                    .equals(checkpointToVerify.getRoot());
                                         requiredCheckpointVerified.set(blockMatches);
                                         return blockMatches;
                                       }))
@@ -327,6 +327,15 @@ public class PeerChainValidator {
         .map(SignedBeaconBlock::getSlot)
         .orElseThrow(
             () -> new IllegalStateException("Missing historical block for slot " + lookupSlot));
+  }
+
+  private boolean validateBlockRootsMatch(
+      final Eth2Peer peer, final Optional<SignedBeaconBlock> mabyeBlock, final Bytes32 root) {
+    if (mabyeBlock.isEmpty()) {
+      LOG.debug("Peer validation failed because it did not provide requested finalized block");
+      return false;
+    }
+    return validateBlockRootsMatch(peer, mabyeBlock.get(), root);
   }
 
   private boolean validateBlockRootsMatch(

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/ResponseStream.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/ResponseStream.java
@@ -13,10 +13,21 @@
 
 package tech.pegasys.teku.networking.eth2.rpc.core;
 
+import java.util.Optional;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
 
 public interface ResponseStream<O> {
-  SafeFuture<O> expectSingleResponse();
+  default SafeFuture<O> expectSingleResponse() {
+    return expectOptionalResponse()
+        .thenApply(
+            maybeResult ->
+                maybeResult.orElseThrow(
+                    () ->
+                        new IllegalStateException(
+                            "No response received when single repsonse expected")));
+  }
+
+  SafeFuture<Optional<O>> expectOptionalResponse();
 
   SafeFuture<Void> expectNoResponse();
 

--- a/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/ResponseStreamImpl.java
+++ b/networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/ResponseStreamImpl.java
@@ -16,6 +16,7 @@ package tech.pegasys.teku.networking.eth2.rpc.core;
 import static com.google.common.base.Preconditions.checkNotNull;
 
 import com.google.common.base.Preconditions;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
@@ -27,7 +28,7 @@ public class ResponseStreamImpl<O> implements ResponseStream<O> {
   private volatile ResponseStreamListener<O> responseListener;
 
   @Override
-  public SafeFuture<O> expectSingleResponse() {
+  public SafeFuture<Optional<O>> expectOptionalResponse() {
     final AtomicReference<O> firstResponse = new AtomicReference<>();
     return expectMultipleResponses(
             response -> {
@@ -37,12 +38,7 @@ public class ResponseStreamImpl<O> implements ResponseStream<O> {
               }
               return SafeFuture.COMPLETE;
             })
-        .thenApply(
-            done -> {
-              checkNotNull(
-                  firstResponse.get(), "No response received when single response expected");
-              return firstResponse.get();
-            });
+        .thenApply(done -> Optional.ofNullable(firstResponse.get()));
   }
 
   @Override

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidatorTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidatorTest.java
@@ -23,7 +23,6 @@ import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_epoc
 import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
 
 import java.util.Optional;
-import java.util.concurrent.CompletionException;
 import org.apache.tuweni.bytes.Bytes32;
 import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem;
 import org.junit.jupiter.api.BeforeEach;
@@ -283,7 +282,7 @@ public class PeerChainValidatorTest {
     forksMatch();
     finalizedCheckpointsMatch();
     when(peer.requestBlockByRoot(requiredCheckpoint.getRoot()))
-        .thenReturn(SafeFuture.failedFuture(new CompletionException(new NullPointerException())));
+        .thenReturn(SafeFuture.completedFuture(Optional.empty()));
 
     final SafeFuture<Boolean> result = peerChainValidator.validate(peer, remoteStatus);
     assertPeerChainRejected(result, DisconnectReason.IRRELEVANT_NETWORK);

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidatorTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/peers/PeerChainValidatorTest.java
@@ -247,9 +247,9 @@ public class PeerChainValidatorTest {
     forksMatch();
     finalizedCheckpointsMatch();
     when(peer.requestBlockByRoot(requiredCheckpoint.getRoot()))
-        .thenReturn(SafeFuture.completedFuture(earlierBlock));
+        .thenReturn(SafeFuture.completedFuture(Optional.of(earlierBlock)));
     when(peer.requestBlockBySlot(earlierBlock.getSlot()))
-        .thenReturn(SafeFuture.completedFuture(earlierBlock));
+        .thenReturn(SafeFuture.completedFuture(Optional.of(earlierBlock)));
 
     final SafeFuture<Boolean> result = peerChainValidator.validate(peer, remoteStatus);
     assertPeerChainVerified(result);
@@ -266,9 +266,9 @@ public class PeerChainValidatorTest {
     final SignedBeaconBlock nonMatchingBlock =
         dataStructureUtil.randomSignedBeaconBlock(earlierCheckpoint.getEpochStartSlot());
     when(peer.requestBlockByRoot(requiredCheckpoint.getRoot()))
-        .thenReturn(SafeFuture.completedFuture(earlierBlock));
+        .thenReturn(SafeFuture.completedFuture(Optional.of(earlierBlock)));
     when(peer.requestBlockBySlot(earlierBlock.getSlot()))
-        .thenReturn(SafeFuture.completedFuture(nonMatchingBlock));
+        .thenReturn(SafeFuture.completedFuture(Optional.of(nonMatchingBlock)));
 
     final SafeFuture<Boolean> result = peerChainValidator.validate(peer, remoteStatus);
     assertPeerChainRejected(result, DisconnectReason.IRRELEVANT_NETWORK);
@@ -370,7 +370,8 @@ public class PeerChainValidatorTest {
   }
 
   private void remoteChainIsAheadOnSameChain() {
-    final SafeFuture<SignedBeaconBlock> blockFuture = SafeFuture.completedFuture(earlierBlock);
+    final SafeFuture<Optional<SignedBeaconBlock>> blockFuture =
+        SafeFuture.completedFuture(Optional.of(earlierBlock));
     final SafeFuture<Optional<SignedBeaconBlock>> optionalBlockFuture =
         SafeFuture.completedFuture(Optional.of(earlierBlock));
 
@@ -394,8 +395,8 @@ public class PeerChainValidatorTest {
   }
 
   private void remoteChainIsAheadOnDifferentChain() {
-    final SafeFuture<SignedBeaconBlock> blockFuture =
-        SafeFuture.completedFuture(randomBlock(earlierBlockSlot));
+    final SafeFuture<Optional<SignedBeaconBlock>> blockFuture =
+        SafeFuture.completedFuture(Optional.of(randomBlock(earlierBlockSlot)));
     final SafeFuture<Optional<SignedBeaconBlock>> optionalBlockFuture =
         SafeFuture.completedFuture(Optional.of(earlierBlock));
 
@@ -406,7 +407,7 @@ public class PeerChainValidatorTest {
   }
 
   private void remoteChainIsAheadAndUnresponsive() {
-    final SafeFuture<SignedBeaconBlock> blockFuture =
+    final SafeFuture<Optional<SignedBeaconBlock>> blockFuture =
         SafeFuture.failedFuture(new NullPointerException());
     final SafeFuture<Optional<SignedBeaconBlock>> optionalBlockFuture =
         SafeFuture.completedFuture(Optional.of(earlierBlock));

--- a/sync/src/main/java/tech/pegasys/teku/sync/gossip/FetchBlockTask.java
+++ b/sync/src/main/java/tech/pegasys/teku/sync/gossip/FetchBlockTask.java
@@ -86,7 +86,11 @@ class FetchBlockTask {
     numberOfRuns.incrementAndGet();
     queriedPeers.add(peer.getId());
     return peer.requestBlockByRoot(blockRoot)
-        .thenApply(FetchBlockResult::createSuccessful)
+        .thenApply(
+            maybeBlock ->
+                maybeBlock
+                    .map(FetchBlockResult::createSuccessful)
+                    .orElseGet(() -> FetchBlockResult.createFailed(Status.FETCH_FAILED)))
         .exceptionally(
             err -> {
               LOG.debug("Failed to fetch block " + blockRoot, err);

--- a/sync/src/test/java/tech/pegasys/teku/sync/gossip/FetchBlockTaskTest.java
+++ b/sync/src/test/java/tech/pegasys/teku/sync/gossip/FetchBlockTaskTest.java
@@ -19,6 +19,7 @@ import static org.mockito.Mockito.when;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 import org.apache.tuweni.bytes.Bytes32;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -50,7 +51,8 @@ public class FetchBlockTaskTest {
     assertThat(task.getBlockRoot()).isEqualTo(blockRoot);
 
     final Eth2Peer peer = registerNewPeer(1);
-    when(peer.requestBlockByRoot(blockRoot)).thenReturn(SafeFuture.completedFuture(block));
+    when(peer.requestBlockByRoot(blockRoot))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(block)));
 
     final SafeFuture<FetchBlockResult> result = task.run();
     assertThat(result).isDone();
@@ -117,7 +119,8 @@ public class FetchBlockTaskTest {
 
     // Add another peer
     final Eth2Peer peer2 = registerNewPeer(2);
-    when(peer2.requestBlockByRoot(blockRoot)).thenReturn(SafeFuture.completedFuture(block));
+    when(peer2.requestBlockByRoot(blockRoot))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(block)));
 
     // Retry
     final SafeFuture<FetchBlockResult> result2 = task.run();
@@ -140,7 +143,8 @@ public class FetchBlockTaskTest {
     when(peer.getOutstandingRequests()).thenReturn(1);
     // Add another peer
     final Eth2Peer peer2 = registerNewPeer(2);
-    when(peer2.requestBlockByRoot(blockRoot)).thenReturn(SafeFuture.completedFuture(block));
+    when(peer2.requestBlockByRoot(blockRoot))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(block)));
     when(peer2.getOutstandingRequests()).thenReturn(0);
 
     // We should choose the peer that is less busy, which successfully returns the block
@@ -158,7 +162,8 @@ public class FetchBlockTaskTest {
     FetchBlockTask task = FetchBlockTask.create(eth2Network, blockRoot);
 
     final Eth2Peer peer = registerNewPeer(1);
-    when(peer.requestBlockByRoot(blockRoot)).thenReturn(SafeFuture.completedFuture(block));
+    when(peer.requestBlockByRoot(blockRoot))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(block)));
 
     task.cancel();
     final SafeFuture<FetchBlockResult> result = task.run();

